src/Orleans.Runtime/Hosting/SiloHostedService.cs

```diff
@@ -0,0 +1,48 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace Orleans.Hosting
+{
+    internal class SiloHostedService : IHostedService
+    {
+        private readonly IServiceProvider serviceProvider;
+        private readonly ILogger logger;
+        private readonly ISiloHost siloHost;
+
+        public SiloHostedService(IServiceProvider serviceProvider)
+        {
+            this.serviceProvider = serviceProvider;
+            this.ValidateSystemConfiguration();
+
+            this.siloHost = serviceProvider.GetRequiredService<ISiloHost>();
+            this.logger = serviceProvider.GetRequiredService<ILoggerFactory>().CreateLogger<SiloHostedService>();
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            this.logger.LogInformation("Starting Orleans Silo.");
```

Comment:
If a hosted service does not start the Host will not fail to start. I think you should catch the exception here and stop the process. See: https://github.com/aspnet/AspNetCore/issues/5900

Reply:
@davidfowl do you have an opinion here? I see that the behavior is changing for 3.0

Reply:
Stop the process? What would that look like?

Reply:
Using IApplicationLifetime and calling StopApplication() ;)

Reply:
I’m not 100% sure that would work well. We don’t exactly plan for people to call Stop inside of StartAsync. I wouldn’t be surprised if it broke in weird ways

Reply:
I'd be ok with either:

* Holding this PR until Hosting propagates `IHostedService.StartAsync` exceptions out through `IHost.StartAsync`
* Providing some way for users to customize the behavior here.

I'd really rather not hard-code behavior to terminate the process abruptly when `StartAsync` fails. That would be painful UX, IMHO.

Or is there a suggested alternative?

Reply:
I don't think we should be stopping the process we don't own. Bubbling an exception up seems like the right and expected behavior here to me.

Reply:
To me, the question is who own the process. If we (the framework) didn't start it, we shouldn't kill it. I think we should signal to the app (with an exception) that we failed to start, and the app can kill the process if it desires so.

Reply:
Note aspnet/AspNetCore#5900 only applies to webhost. The generic host already propagates StartAsync exceptions. Since we're moving AspNetCore to use generic host in 3.0 anyways I don't see any reason to block on a bug in webhost.

Reply:
Thanks, @Tratcher, I didn't realize that. Given that this, I suppose we should consider this a non-issue (or a known issue with webhost).

Reply:
But please notify the host then, I would like to use this and need to stop the process. I had an issue with that because right now I host the silo in an hosted service and first it looked like the deployment went fine and then everything blew up later.

Reply:
@SebastianStehle can you use `HostBuilder` instead of `WebHostBuilder`?

`WebHostBuilder` (really [HostedServiceExecutor](https://github.com/aspnet/AspNetCore/blob/1aa50faa290ecda304507981cd01ed92651d5e34/src/Hosting/Hosting/src/Internal/HostedServiceExecutor.cs#L30)) has this issue but `HostBuilder` does not. `WebHostBuilder` will be fixed in 3.0.

You could work around `WebHostBuilder`'s implementation by manually removing `SiloHostedService`, eg:

```C#
var defaultSiloHostedService = services.FirstOrDefault(svc => svc.ImplementationType.Name == "SiloHostedService");
                    if (defaultSiloHostedService != null) services.Remove(defaultSiloHostedService);
```

Then add your own `IHostedService` based on the existing `SiloHostedService` (nothing internal in there).

Would that work?

The alternative is that we introduce some mechanism for you to customize the behavior but I feel that's adding API/code where there shouldn't be any.

Reply:
Sure, that should work, I was not aware that HostBuilder does not have this problem.